### PR TITLE
Enable intergroup references in "advanced usage" mode

### PR DESCRIPTION
**Pull request by tpdownes**
- disable errors when references across groups are made
- add "advanced usage" warning to instructions for groups that make reference to previous groups
- modify metadata to eliminate "inputs" field and replace with the graph connections from the group to "deployment_inputs" and "intergroup_inputs"; these are the deployment variables and intergroup outputs necessary to resolve the group input settings (in Terraform these match group input variable names, for Packer they may not)
- enable deep resolution of deployment variables in Packer settings that are structural types
    - i.e. one can embed the value of `$(vars.zone)` in a object or tuple in `default.auto.pkrvars.hcl`
- eliminate "hidden" 2-field use keyword in favor of single field containing only the module

Packer groups are still subject to further development based upon active work building upon #1180. A multi-group Terraform deployment is shown below. If one executes the following command between invocations of `terraform apply` it will propagate the values to the second group. We anticipate automating this functionality in the near future.

```shell
terraform -chdir=igc-tf-test/zero output -json | jq 'with_entries(.value = .value.value)' > igc-tf-test/two/inputs.auto.tfvars.json
```

```yaml
---
blueprint_name: igc

vars:
  project_id: toolkit-demo-zero-e913
  deployment_name: igc-tf-test
  region: us-east4
  zone: us-east4-c

deployment_groups:
- group: zero
  modules:
  - id: network0
    source: modules/network/vpc
  - id: homefs
    source: modules/file-system/filestore
    use: [network0]
    settings:
      local_mount: /home
  - id: projectsfs
    source: modules/file-system/filestore
    use: [network0]
    settings:
      local_mount: /projects
- group: two
  modules:
  - id: network1
    source: modules/network/pre-existing-vpc
    use:
    - network0
  - id: vm
    source: modules/compute/vm-instance
    use:
    - network1
    - homefs
    - projectsfs
    settings:
      machine_type: n2-standard-4
```

A Packer blueprint is shown below. This has more limited functionality due to variable name matching and evaluation of settings that we have discussed offline. We anticipate automating both of these issues in near future.

```yaml
---
blueprint_name: igc

vars:
  project_id: toolkit-demo-zero-e913
  deployment_name: igc-pkr-test
  region: us-east4
  zone: us-east4-c

deployment_groups:
- group: zero
  modules:
  - id: network0
    source: modules/network/vpc
  - id: homefs
    source: modules/file-system/filestore
    use: [network0]
    settings:
      local_mount: /home
  - id: projectsfs
    source: modules/file-system/filestore
    use: [network0]
    settings:
      local_mount: /projects
  - id: script
    source: modules/scripts/startup-script
    settings:
      runners:
      - type: shell
        destination: hellow.sh
        content: |
          #!/bin/bash
          echo "Hello, World!"

- group: one
  modules:
  - id: image
    source: modules/packer/custom-image
    kind: packer
    use:
    - network0
    - script

- group: two
  modules:
  - id: network1
    source: modules/network/pre-existing-vpc
    use:
    - network0
  - id: low_cost_ng
    source: community/modules/compute/schedmd-slurm-gcp-v5-node-group
    settings:
      node_count_dynamic_max: 10
      machine_type: n2-standard-4

  - id: low_cost_partition
    source: community/modules/compute/schedmd-slurm-gcp-v5-partition
    use:
    - network1
    - homefs
    - projectsfs
    - low_cost_ng
    settings:
      partition_name: low_cost
```

### Submission Checklist

* [x] Have you installed and run this change against pre-commit? (`pre-commit
  install`)
* [x] Are all tests passing? (`make tests`)
* [x] Have you written unit tests to cover this change?
* [x] Is unit test coverage still above 80%?
* [x] Have you updated all applicable documentation?
* [x] Have you followed the guidelines in our Contributing document?

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -113,6 +111,27 @@ func (dc DeploymentConfig) getGroupByID(groupID string) (DeploymentGroup, error)
 	return group, nil
 }
 
+// ModuleGroup returns the group ID containing the module
+func (b Blueprint) ModuleGroup(mod string) (string, error) {
```
Comment (mr0re1) [inline pkg/config/config.go:115]: Let's return whole `DeploymentGroup`

File: pkg/config/config.go [hunk 2]
```diff
@@ -113,6 +111,27 @@ func (dc DeploymentConfig) getGroupByID(groupID string) (DeploymentGroup, error)
...
+				return g.Name, nil
+			}
+		}
+	}
+	return "", fmt.Errorf("%s: %s", errorMessages["invalidMod"], mod)
+}
+
+// ModuleGroupOrDie returns the group ID containing the module; panics if unfound
+func (b Blueprint) ModuleGroupOrDie(mod string) string {
+	modID, err := b.ModuleGroup(mod)
```
Comment (mr0re1) [inline pkg/config/config.go:128]: `modID` -> `g`

File: pkg/config/config.go [hunk 3]
```diff
@@ -303,7 +322,23 @@ func (c ModConnection) IsUseKind() bool {
 
 // GetSharedVariables returns variables used in the connection (can be empty!)
 func (c ModConnection) GetSharedVariables() []string {
-	return c.sharedVariables
+	var vars []string
```
Comment (mr0re1) [inline pkg/config/config.go:325]: nit. Readability can be improved:
```go
if !c.ref.IsIntergroup() {
  return c.sharedVariables
}
...else  do AutomaticOutputName for each
```

Comment (tpdownes) [reply pkg/config/config.go:325]: Done

File: pkg/config/config.go (hunk 1 above)
Comment (tpdownes) [reply pkg/config/config.go:115]: Done

File: pkg/config/config.go [hunk 4]
```diff
@@ -799,35 +828,67 @@ func ConvertMapToCty(iMap map[string]interface{}) (map[string]cty.Value, error)
...
-					return diags
-				}
-				ctyMap[key] = newVal
-			}
+		newVal, err := cty.Transform(val, func(p cty.Path, v cty.Value) (cty.Value, error) {
+			return resolveValue(v, evalCtx, unknownMap)
+		})
+
+		var re *ResolutionError
+		if errors.As(err, &re) && re.allowed {
```
Comment (mr0re1) [inline pkg/config/config.go:842]: I think it's overly complex way to do it.
Could we instead do something like this without leaking to `ResolveVariables`?:
```go
// packerwtiter
dependOnIgc := findSettingsDependingOnIgc(mod.Settings.Items()) // set of setting names
defaults := removeKeys(mod.Settings.Items(), dependOnIgc) 
defaults = ResolveVariables(defaults, dc.Config.Vars.Items())
writeautovars(defaults)
printPackerInstructions(modPath, mod.ID,  len(dependOnIgc) > 0)
```

**Review by mr0re1 — APPROVED** [review]

File: pkg/config/config.go (hunk 4 above)
Comment (tpdownes) [reply pkg/config/config.go:842]: We will implement this alongside other planned changes in very near-term.

**Pull request merged**